File: python/graphs/surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        neighbors = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        starts = []
        for i in range(len(board)):
            for j in range(len(board[0])):
                if i == 0 or j == 0 or i == len(board) - 1 or j == len(board[0]) - 1:
                    starts.append([i, j])

        def dfs(pos):
            if pos[0] < 0 or pos[1] < 0 or pos[0] >= len(board) or pos[1] >= len(board[0]) or board[pos[0]][
                pos[1]] != "O":
                return

            board[pos[0]][pos[1]] = "."

            for n in neighbors:
                dfs([pos[0] + n[0], pos[1] + n[1]])

        for pos in starts:
            dfs(pos)

        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == "O":
                    board[i][j] = "X"
                if board[i][j] == ".":
                    board[i][j] = "O"

File: c++/graphs/surrounded-regions.cc
class Solution {
public:
    void solve(vector<vector<char>>& board) {
        std::vector<std::pair<int, int>> neighbors = {std::make_pair(0, 1), std::make_pair(1, 0), std::make_pair(-1, 0), std::make_pair(0, -1)};
        for (int i = 0; i < board.size(); ++i) {
            for (int j = 0; j < board[0].size(); ++j) {
                if (i == 0 || j == 0 || i == board.size() - 1 || j == board[0].size() - 1) {
                    dfs(i, j, board, neighbors);
                }
            }
        }

        for (int i = 0; i < board.size(); ++i) {
            for (int j = 0; j < board[0].size(); ++j) {
                if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
                if (board[i][j] == '.') {
                    board[i][j] = 'O';
                }
            }
        }
    }

private:
    void dfs(int i, int j, std::vector<std::vector<char>>& board, std::vector<std::pair<int, int>>& neighbors) {
        if (i < 0 || j < 0 || i >= board.size() || j >= board[0].size() || board[i][j] != 'O') {
            return;
        }

        board[i][j] = '.';

        for (auto [n_i, n_j] : neighbors) {
            dfs(i+n_i, j+n_j, board, neighbors);
        }
    }
};
